contrib_bots/lib/giphy/giphy.py

```diff
@@ -0,0 +1,140 @@
+# To use this plugin, you need to set up the Giphy API key for this bot in
+# ~/.giphy_config
+
+from __future__ import absolute_import
+from __future__ import print_function
+import six
+import requests
+import logging
+import sys
+import os
+import re
+if six.PY2:
+    from six.moves.configparser import SafeConfigParser
+else:
+    from configparser import SafeConfigParser
+
+GIPHY_TRANSLATE_API = 'http://api.giphy.com/v1/gifs/translate'
+
+if not os.path.exists(os.environ['HOME'] + '/.giphy_config'):
+    print('Giphy bot config file not found, please set up it in ~/.giphy_config'
+          '\n\nUsing format:\n\n[giphy-config]\nkey=<giphy API key here>\n\n')
+    sys.exit(1)
+
+
+class GiphyHandler(object):
+    '''
+    This plugin posts a GIF in response to the keywords provided by the user.
+    Images are provided by Giphy, through the public API.
+    The bot looks for messages starting with "@giphy" or @mention of the bot
+    and responds with a message with the GIF based on provided keywords.
+    It also responds to private messages.
+    '''
+    def usage(self):
+        return '''
+            This plugin allows users to post GIFs provided by Giphy.
+            Users should preface keywords with "@giphy" or the Giphy-bot @mention.
+            The bot responds also to private messages.
+            '''
+
+    def triage_message(self, message, client):
+        # To prevent infinite loop in private message, bot will detect
+        # if the sender name is the bot name it will return false.
+        if message['type'] == 'private':
+            return client.full_name != message['sender_full_name']
+
+        # Return True if we want to (possibly) response to this message.
+        original_content = message['content']
+        is_giphy_called = (original_content.startswith('@giphy ') or
+                           message['is_mentioned'])
+
+        return is_giphy_called
+
+    def handle_message(self, message, client, state_handler):
+        bot_response = get_bot_giphy_response(message, client)
+
+        if message['type'] == 'private':
+            client.send_message(dict(
+                type='private',
+                to=message['sender_email'],
+                content=bot_response,
+            ))
+        else:
+            client.send_message(dict(
+                type='stream',
+                to=message['display_recipient'],
+                subject=message['subject'],
+                content=bot_response,
+            ))
+
+
+class GiphyNoResultException(Exception):
+    pass
+
+
+def get_giphy_api_key_from_config():
+    config = SafeConfigParser()
+    with open(os.environ['HOME'] + '/.giphy_config', 'r') as config_file:
+        if six.PY2:
```

Comment:
OK, this doesn't look that good but I realize it is unavoidable. Wish I have tested it myself before suggesting it. I think be can ignore the deprecation of `readfp` for now. Or we can go with this one to be more correct, it's your choice.